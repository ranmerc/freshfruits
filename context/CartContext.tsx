import { useLocalStorage } from "usehooks-ts";
import {
  createContext,
  useContext,
  ReactNode,
  useReducer,
  Dispatch,
  useCallback,
} from "react";
import CartItem from "@/types/CartItem";

type ActionType =
  | {
      type: "ADD_TO_CART";
      payload: CartItem;
    }
  | {
      type: "REMOVE_FROM_CART";
      payload: {
        fruitId: number;
        selectedPackId: number;
      };
    }
  | {
      type: "CLEAR_CART";
    };

interface CartContextType {
  cartItems: CartItem[];
  cartDispatch: Dispatch<ActionType>;
}

const CartContext = createContext<CartContextType | undefined>(undefined);

export function useCartContext() {
  const context = useContext(CartContext);

  if (!context) {
    throw new Error("useCartContext must be used within a CartProvider");
  }

  return context;
}

function cartReducer(state: CartItem[], action: ActionType) {
  switch (action.type) {
    case "ADD_TO_CART": {
      const { fruitId, selectedPackId, quantity } = action.payload;

      const existingItemIndex = state.findIndex(
        (item) =>
          item.fruitId === fruitId && item.selectedPackId === selectedPackId
      );

      if (existingItemIndex !== -1) {
        const newState = [...state];
        newState[existingItemIndex].quantity += quantity;
        return newState;
      }

      return [...state, action.payload];
    }

    case "REMOVE_FROM_CART": {
      const { fruitId, selectedPackId } = action.payload;

      const existingItemIndex = state.findIndex(
        (item) =>
          item.fruitId === fruitId && item.selectedPackId === selectedPackId
      );

      if (existingItemIndex === -1) {
        return state;
      }

      const newState = [...state];
      newState[existingItemIndex].quantity -= 1;

      if (newState[existingItemIndex].quantity === 0) {
        newState.splice(existingItemIndex, 1);
      }

      return newState;
    }

    case "CLEAR_CART":
      return [];
    default:
      return state;
  }
}

export function CartProvider({ children }: { children: ReactNode }) {
  const [savedCartItems, setSavedCartItems] = useLocalStorage<CartItem[]>(
    "cart-items",
    []
  );

  const reducerLocalStorage = useCallback(
    (state: CartItem[], action: ActionType) => {
      const newState = cartReducer(state, action);

      setSavedCartItems(newState);

      return newState;
    },
    [setSavedCartItems]
  );

  const [state, dispatch] = useReducer(reducerLocalStorage, savedCartItems);

  return (
    <CartContext.Provider value={{ cartItems: state, cartDispatch: dispatch }}>
      {children}
    </CartContext.Provider>
  );
}
