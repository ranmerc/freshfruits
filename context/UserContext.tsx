import { createContext, useContext, ReactNode } from "react";
import { useLocalStorage } from "usehooks-ts";

export interface UserContextType {
  username: string;
  setUserName: (user: string) => void;
}

export const UserContext = createContext<UserContextType | undefined>(
  undefined
);

export function useUserContext() {
  const context = useContext(UserContext);

  if (!context) {
    throw new Error("useUserContext must be used within a UserProvider");
  }

  return context;
}

export function UserProvider({ children }: { children: ReactNode }) {
  const [username, setUserName] = useLocalStorage<string>("username", "");

  return (
    <UserContext.Provider value={{ username, setUserName }}>
      {children}
    </UserContext.Provider>
  );
}
