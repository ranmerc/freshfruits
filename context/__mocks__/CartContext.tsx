export const CartContext = jest.fn();

export const useCartContext = jest.fn();

export const CartProvider = jest.fn();
