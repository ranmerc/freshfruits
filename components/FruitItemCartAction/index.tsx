import { useCartContext } from "@/context/CartContext";
import PriceData from "@/types/PriceData";
import { AddCircle, RemoveCircle } from "@mui/icons-material";
import {
  Button,
  FormControl,
  IconButton,
  InputLabel,
  MenuItem,
  Select,
  Stack,
  Typography,
} from "@mui/material";
import React, { useMemo } from "react";
import calculateDiscount from "@/utils/calculateDiscount";

export default function FruitItemCartAction({
  prices,
  id,
}: {
  prices: PriceData;
  id: number;
}) {
  const [selectedPack, setSelectedPack] = React.useState(0);
  const { cartItems, addItemToCart, removeItemFromCart } = useCartContext();

  const quantityInCart = useMemo(() => {
    const item = cartItems.find(
      (item) => item.fruitId === id && item.selectedPackId === selectedPack
    );
    return item?.quantity || 0;
  }, [cartItems, id, selectedPack]);

  const itemToBeAdded = useMemo(() => {
    return {
      fruitId: id,
      pack: prices.packs[selectedPack],
      quantity: 1,
      selectedPackId: selectedPack,
    };
  }, [id, selectedPack, prices]);

  return (
    <>
      <Stack rowGap={2}>
        {prices.packs[selectedPack].price !==
          prices.packs[selectedPack].discountPrice && (
          <Typography
            variant="caption"
            sx={{
              bgcolor: "secondary.main",
              padding: 0.5,
              fontSize: "0.8rem",
              textAlign: "center",
            }}
          >
            {calculateDiscount(
              prices.packs[selectedPack].discountPrice,
              prices.packs[selectedPack].price
            )}{" "}
            % off
          </Typography>
        )}
        <FormControl fullWidth>
          <InputLabel id={`packs-for-${id}`}>Pack</InputLabel>
          <Select
            labelId={`packs-for-${id}`}
            id={`select-packs-for-${id}`}
            value={selectedPack}
            label="Pack"
            onChange={(e) => {
              setSelectedPack(e.target.value as number);
            }}
          >
            {prices.packs.map((pack, i) => {
              if (pack.inStock) {
                return (
                  <MenuItem value={i} key={i}>
                    {pack.quantity} {pack.type === "count" ? "Nos" : "gm"} for â‚¹
                    {pack.discountPrice}
                  </MenuItem>
                );
              }
            })}
          </Select>
        </FormControl>

        {quantityInCart !== 0 ? (
          <Stack
            direction={"row"}
            alignItems={"center"}
            sx={{
              bgcolor: "secondary.main",
              borderRadius: 2,
              justifyContent: "space-around",
            }}
          >
            <IconButton
              onClick={() =>
                removeItemFromCart({
                  fruitId: id,
                  selectedPackId: selectedPack,
                })
              }
              aria-label="Remove from Cart"
            >
              <RemoveCircle />
            </IconButton>
            <Typography fontSize={"1.2rem"}>{quantityInCart}</Typography>
            <IconButton
              onClick={() => addItemToCart(itemToBeAdded)}
              aria-label="Add to Cart"
            >
              <AddCircle />
            </IconButton>
          </Stack>
        ) : (
          <Button
            size="large"
            variant="outlined"
            onClick={() => addItemToCart(itemToBeAdded)}
          >
            Add
          </Button>
        )}
      </Stack>
    </>
  );
}
