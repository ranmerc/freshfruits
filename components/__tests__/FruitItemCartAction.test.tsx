import { screen, render } from "@testing-library/react";
import "@testing-library/jest-dom";
import FruitItemCartAction from "../FruitItemCartAction";
import { useCartContext } from "@/context/CartContext";

jest.mock("@/context/CartContext");

const mockedUseCartContext = useCartContext as jest.Mock;

describe("FruitItemCartAction component", () => {
  it("Renders correct price details - 1", () => {
    mockedUseCartContext.mockReturnValue({
      cartItems: [],
      addItemToCart: jest.fn(),
      removeItemFromCart: jest.fn(),
    });

    render(
      <FruitItemCartAction
        prices={{
          id: 1,
          packs: [
            {
              type: "count",
              quantity: 1,
              price: 100,
              discountPrice: 90,
              inStock: true,
            },
            {
              type: "count",
              quantity: 6,
              price: 400,
              discountPrice: 300,
              inStock: false,
            },
          ],
        }}
        id={1}
      />
    );

    // Discount percentage
    expect(screen.getByText(/10 % off/i)).toBeInTheDocument();

    expect(screen.getByText(/1 Nos for ₹90/i)).toBeInTheDocument();

    expect(screen.getByRole("button", { name: "Add" })).toBeInTheDocument();
  });

  it("Renders correct price details - 2", () => {
    mockedUseCartContext.mockReturnValue({
      cartItems: [
        {
          fruitId: 1,
          selectedPackId: 0,
          pack: {
            type: "count",
            quantity: 6,
            price: 400,
            discountPrice: 300,
            inStock: true,
          },
          quantity: 4,
        },
      ],
      addItemToCart: jest.fn(),
      removeItemFromCart: jest.fn(),
    });

    render(
      <FruitItemCartAction
        prices={{
          id: 1,
          packs: [
            {
              type: "count",
              quantity: 6,
              price: 400,
              discountPrice: 300,
              inStock: true,
            },
            {
              type: "count",
              quantity: 1,
              price: 100,
              discountPrice: 90,
              inStock: true,
            },
          ],
        }}
        id={1}
      />
    );

    expect(screen.getByText(/25 % off/i)).toBeInTheDocument();

    expect(screen.getByText(/6 Nos for ₹300/i)).toBeInTheDocument();

    expect(
      screen.queryByRole("button", { name: "Add" })
    ).not.toBeInTheDocument();
  });
});
