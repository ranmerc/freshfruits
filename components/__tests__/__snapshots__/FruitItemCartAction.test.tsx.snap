// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FruitItemCartAction component Renders correct price details - 1 1`] = `
<div
  className="MuiStack-root css-pwukzo-MuiStack-root"
>
  <span
    className="MuiTypography-root MuiTypography-caption css-1heknby-MuiTypography-root"
  >
    10
     
    % off
  </span>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      id="packs-for-1"
    >
      Pack
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      onClick={[Function]}
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="packs-for-1 select-packs-for-1"
        className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="select-packs-for-1"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        1
         
        Nos
         for ₹
        90
      </div>
      <input
        aria-hidden={true}
        aria-invalid={false}
        className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        disabled={false}
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value={0}
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-14lo706"
        >
          <span>
            Pack
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge css-9y1egq-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Add
  </button>
</div>
`;

exports[`FruitItemCartAction component Renders correct price details - 2 1`] = `
<div
  className="MuiStack-root css-pwukzo-MuiStack-root"
>
  <span
    className="MuiTypography-root MuiTypography-caption css-1heknby-MuiTypography-root"
  >
    25
     
    % off
  </span>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      id="packs-for-1"
    >
      Pack
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      onClick={[Function]}
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="packs-for-1 select-packs-for-1"
        className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="select-packs-for-1"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        6
         
        Nos
         for ₹
        300
      </div>
      <input
        aria-hidden={true}
        aria-invalid={false}
        className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        disabled={false}
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value={0}
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-14lo706"
        >
          <span>
            Pack
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiStack-root css-i42th-MuiStack-root"
  >
    <button
      aria-label="Remove one item"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="RemoveCircleIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11H7v-2h10v2z"
        />
      </svg>
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <p
      className="MuiTypography-root MuiTypography-body1 css-b02fjb-MuiTypography-root"
    >
      4
    </p>
    <button
      aria-label="Add one item"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddCircleIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
        />
      </svg>
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
